### PR TITLE
Extend documentation about test assertions, archetype cosmetics

### PR DESCRIPTION
**Pull request by htreu**
Addresses the discussion about Hamcrest & JUnit assertions started here: https://github.com/eclipse/smarthome/pull/6118#issuecomment-416131766

Conclusion: Hamcrest assertions should be used, JUnit only for basic checks.

Also addresses this review comment in the archetype: https://github.com/eclipse/smarthome/pull/6000#discussion_r212791579

Signed-off-by: Henning Treu <henning.treu@telekom.de>

### CODE REVIEW

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 1]
```diff
@@ -61,20 +61,23 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
 
     @Override
     public void initialize() {
+        // logger.debug("Start initializing!");
         config = getConfigAs(${bindingIdCamelCase}Configuration.class);
 
         // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
-        // Long running initialization should be done asynchronously in background.
+        // Long running initialization (e.g. WAN access) should be done asynchronously in the background. Set a status 
+        // of UNKNOWN or OFFLINE in these situations and return quickly. Set the status to ONLINE once a proper
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: I think it does not become clear here that if you do async init, the status when returning from this method should be set to UNKNOWN (do you see any valid usecase for OFFLINE as well?) and the async code must then change it to ONLINE or OFFLINE asap.
Especially line 81 below can probably easily confuse people what to do here.

Comment (htreu) [reply tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: I agree and admit that I had a hard time even for this small sentence ;-)
I will rephrase the whole paragraph.

Comment (htreu) [reply tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: done.

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 2]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        // can decide it directly.
+        // In case you can not decide the thing status directly (e.g. for long running initialization using WAN access
+        // or similar) you should set status OFFLINE here and then decide the real status asynchronously in the
+        // background.
 
-        // logger.debug("Start initializing!");
-        updateStatus(ThingStatus.ONLINE);
+        // Example for background initialization:
+        scheduler.execute(() -> {
+            boolean initialized = true; // <background task with long running initialization here>
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: `initialized` is misleading here - it should rather be `reachable` or `online` here as this is what you are trying to figure out.

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 3]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
         config = getConfigAs(${bindingIdCamelCase}Configuration.class);
 
-        // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
-        // Long running initialization should be done asynchronously in background.
+        // TODO: Initialize the thing.
+        // The framework requires you to leave this method quickly. Also, before leaving this method a thing status
+        // from one of ONLINE, OFFLINE or UNKNOWN must be set. This might already be the real thing status in case you
+        // can decide it directly.
+        // In case you can not decide the thing status directly (e.g. for long running initialization using WAN access
+        // or similar) you should set status OFFLINE here and then decide the real status asynchronously in the
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: Don't you mean UNKNOWN here?

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 4]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+            boolean initialized = true; // <background task with long running initialization here>
+            // when done do:
+            if (initialized) {
+                // updateStatus(ThingStatus.ONLINE); // commented for stable test execution
+            } else {
+                // updateStatus(ThingStatus.OFFLINE);
+            }
+        });
+
+        // set the thing status to OFFLINE temporarily and let the background task decide for the real status.
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: UNKNOWN

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 5]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        // the framework is then able to reuse the resources from the thing handler initialization.
+        updateStatus(ThingStatus.OFFLINE);
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: UNKNOWN

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 6]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
 
     @Override
     public void initialize() {
+        // logger.debug("Start initializing!");
         config = getConfigAs(${bindingIdCamelCase}Configuration.class);
 
-        // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
-        // Long running initialization should be done asynchronously in background.
+        // TODO: Initialize the thing.
+        // The framework requires you to leave this method quickly. Also, before leaving this method a thing status
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: better: to return from this method quickly

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 7]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        // from one of ONLINE, OFFLINE or UNKNOWN must be set. This might already be the real thing status in case you
+        // can decide it directly.
+        // In case you can not decide the thing status directly (e.g. for long running initialization using WAN access
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: better: for long running connection handshake using WAN

File: tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java [hunk 8]
```diff
@@ -61,23 +61,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
     }
 
     @Override
     public void initialize() {
+        // logger.debug("Start initializing!");
         config = getConfigAs(${bindingIdCamelCase}Configuration.class);
 
-        // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
-        // Long running initialization should be done asynchronously in background.
+        // TODO: Initialize the thing.
```
Comment (kaikreuzer) [inline tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java]: Actually, it is the handler that get's initialised.

**Comment by htreu** [comment]
done.

**Pull request merged**